### ISSUE TITLE
Reduce Loss

### ISSUE THREAD

**Issue by agemagician** [cybertronai/transformer-xl#14]
Hi,

Thanks for the awesome script.

I have made a modification, to get the average of loss, ppl across all machines.
Currently, your script only store the best value for the validation loss only on the master machine. It ignores the validation loss on the rest of the machines.
A simple solution that I made is to call the following function after calculation training and validation loss:

```
def reduce_tensor(value):
    tensor = torch.Tensor([value]).type(torch.cuda.FloatTensor)
    rt = tensor.clone()
    dist.all_reduce(rt, op=dist.reduce_op.SUM)
    rt /= dist.get_world_size()
    return rt.item()
```

of course, you have to change both train and eval function to:

```
            # compute average loss over last logging interval
            cur_loss = train_loss / elapsed_steps
            if dist.get_world_size() > 1:
                reduced_cur_loss = reduce_tensor(cur_loss)
            else:
                reduced_cur_loss = cur_loss
```

and

```
    # Log all the things.
    mean_loss = total_loss / total_len
    if dist.get_world_size() > 1:
        reduced_mean_loss = reduce_tensor(mean_loss)
    else:
        reduced_mean_loss = mean_loss
```

**Comment by yaroslavvb** [comment]
Oh interesting, I assumed `loss = loss.mean()` automatically did this when model was wrapped in DDP. But now looking at DistributedDataParallel code it looks like it's not that smart.
Right now development is split over "dev" and "checkpoint" branches, I'm going to incorporate your suggestion as I do the merge later this week, thanks for the catch!

**Comment by agemagician** [comment]
No worries, you are welcome. 
But, I think this means you need to re-evaluate your 128 GPU model.
Because the reported results use only part of the validation dataset, which exists in the main node.

**Comment by yaroslavvb** [comment]
That mean-loss calculation is just for visual checking in tensorboard, accuracy results come from [evaluate_and_log](https://github.com/cybertronai/transformer-xl/blob/d6333bdbf5a5a72c9105a37cb2e6b189dc0651bf/train.py#L391) function which does single-GPU evaluation

**Comment by agemagician** [comment]
Aha, great :)
